### PR TITLE
Maksimchepelev/feature decorator service

### PR DESCRIPTION
**Pull request by MaksimChepelev**

### CODE REVIEW

**Comment by MaksimChepelev** [comment]
Merge after db pull request

File: src/decorators/authorizationDecorator.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    let req = fArgs[0];
+    let token = req.headers.authorization;
+
+    let user = await UserManager.findByToken(token);
+
+    if (user) {
+      return await f.apply(this, arguments);
+    } else {
+      logger.error('token ' + token + ' incorrect');
+      throw 'authorization failed'
```
Comment (baghbidi) [inline src/decorators/authorizationDecorator.js]: create exception class and throw an instance of that class

File: src/decorators/authorizationDecorator.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    logger.info('checking user token: ' + token);
+
+    let fArgs = Array.prototype.slice.call(arguments);
+    let req = fArgs[0];
+    let token = req.headers.authorization;
+
+    let user = await UserManager.findByToken(token);
+
+    if (user) {
```
Comment (baghbidi) [inline src/decorators/authorizationDecorator.js]: to be more performant:
```
    if (!user) {
      logger.error('token ' + token + ' incorrect')
      throw new NotAuthorized()
    }

    return await f.apply(this, arguments)
```

File: src/decorators/authorizationDecorator.js [hunk 3]
```diff
@@ -10,29 +10,37 @@
...
-    let user = await UserManager.findByToken(token);
+    const user = await UserManager.findByAccessToken(token);
 
-    if (user) {
-      return await f.apply(this, arguments);
-    } else {
+    if (!user) {
       logger.error('token ' + token + ' incorrect');
-      throw 'authorization failed'
+      throw new Errors.AuthorizationError('authorization failed');
```
Comment (baghbidi) [inline src/decorators/authorizationDecorator.js]: this should be `AuthenticationError`, sorry if my comment confused you!
`Authorization` errors should be thrown when an authenticated user tries to access other users' resources. like `user1` has a valid token, but calls the API to modify a microservice that belongs to `user2`.

File: src/decorators/authorizationDecorator.js [hunk 4]
```diff
@@ -10,29 +10,37 @@
...
     }
+    if (Date.now() > user.accessToken.expirationTime) {
+      logger.error('token ' + token + ' expired');
+      throw new Errors.AuthorizationError('token expired');
```
Comment (baghbidi) [inline src/decorators/authorizationDecorator.js]: same here

File: src/decorators/authorizationDecorator.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    const req = fArgs[0];
+    const token = req.headers.authorization;
+
+    logger.info('checking user token: ' + token);
+
+    const user = await UserManager.findByAccessToken(token);
+
+    if (!user) {
+      logger.error('token ' + token + ' incorrect');
+      throw new Errors.AuthenticationError('authorization failed');
```
Comment (baghbidi) [inline src/decorators/authorizationDecorator.js]: `authentication failed`

File: src/decorators/authorizationDecorator.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ *  *******************************************************************************
+ *
+ */
+const logger = require('../logger');
+const config = require('../config');
+const UserManager = require('../sequelize/managers/userManager');
+const AccessTokenManager = require('../sequelize/managers/accessTokenManager');
+const Errors = require('../utils/errors');
+
+function checkAuthorization(f) {
```
Comment (baghbidi) [inline src/decorators/authorizationDecorator.js]: `checkAuthToken`

**Pull request merged**